### ISSUE TITLE
Rootana production 2

### ISSUE THREAD

**Issue by benkrikler** [alcap-org/AlcapDAQ#166]
This is just to get the ball rolling with production 2.  I believe the code is all there now for part 1 of elog:152 but still needs refinement in terms of plot scales and axes and some algorithm parameters.  

If part 2 is going to take some time I will try to write a new AP generator that combines the 3 simple existing ones (max bin, CFT and Simple integral) so we have TAPs with all fields filled.

**Comment by benkrikler** [comment]
Where have we got to with all this then?  I'm planning on running over Al50b tonight.  I think we can only use the run scripts if we've tagged the code, so what's realistic to be ready by tonight (ie. about 7 hours time for me)?

**Comment by AndrewEdmonds11** [comment]
So we have the TAP generator with all fields filled. I'm having some problem with adding the PulseCandidateFinder (the channel names are going a bit screwy so I can't get the parameter values for start/stop conditions) so it might not be done (not helped by the fact that I have to leave early and that my laptop has broken so can't do any work at home) but this isn't a critical problem and we can run without it.

**Comment by benkrikler** [comment]
Is this another problem with the Channel IDs (please god, no..)?  Do you want me to look into something and is this something I could try helping out with if you don't get it done in time?

**Comment by AndrewEdmonds11** [comment]
No worries, just solved it. I was creating the PCF in the constructor of the generator and was passing the channel. But the channel doesn't exist until the `ProcessPulses` function, so it's all OK now.

I may still need to make some tweaks to make sure that the start/stop conditions are good and give us the whole pulse for analysis, though

**Comment by benkrikler** [comment]
Oh, sweet!  If we run tonight and uncover bugs, then hopefully there's still time tomorrow to run again and get results that can be discussed and shown on Wednesday.  Then there's two more days to repeat once more before the Friday meeting.

**Comment by benkrikler** [comment]
On my side, TDPs seem to be working and I'm able to plot the correlated amplitudes, timing difference and plot amplitudes of one channel based on a cut on the other.  Minor tweaks to the output are still going in, and I might break the code up into 2 separate modules, but a first iteration of the code for Task 1 is there then.  A possible extension is to look at the integrals as well as the amplitudes; is this worth doing at this point?

I'm also adding a `PlotIntegrals` module which just does 1D histograms of all Integrals the same as `PlotAmplitudes`.

**Comment by benkrikler** [comment]
The `PlotIntegrals` module is up and the TDP plots for task 1.c and 1.d is in place.

**Comment by benkrikler** [comment]
Nearly everything we need is there now, so I've started testing actual production configs.  

Sadly, It looks like I'm going to have to take the pulse finder out of the MegaTAPs for now since it slows a single event run down from 0.5 seconds to 27.  At half a minute per event it would take 12 hours or so to do a single run, so it's not realistic for Wednesday.  Depending how much we can optimize it before Friday, we may or may not be able to use it.

@AndrewEdmonds11 or @litchfld, would you be happy to start testing a production?  It's 3am for me so I should head to bed.  I started preparing things on the new branch `feature/BK_prepare_production_2`.  The production.cfg modules file has been moved into the configurations directory and updated with everything I would like to run ideally (except Damien's TDiff modules which we'll add once they're ready).  Depending on memory consumption and time we may need to break the runs into smaller steps so we'd need a single run script and 2 or more modules files for each step.

**Comment by AndrewEdmonds11** [comment]
Yes, I can start testing production. I'll update this issue with what I find

**Comment by AndrewEdmonds11** [comment]
I think the massive slow down with the PCF is because it is instantiated in the constructor, where the channel is set to "*", which means the parameter value is 0 and so we get loads O(10) sub pulses.

I've (very) poorly explained this before here:

> No worries, just solved it. I was creating the PCF in the constructor of the generator and was passing the channel. But the channel doesn't exist until the ProcessPulses function, so it's all OK now.

and here:

> Yes, make any changes you like (especially where I've been an idiot). For the initializations in the constructor when I had the PCF there I was getting "*" for the channel string.

Anyway, I know that it might be worse to keep on creating a PCF every TPI but otherwise it doesn't work. Unless I find a way to have the channel correctly defined in the constructor...

**Comment by AndrewEdmonds11** [comment]
So it makes no real difference in terms of how long it takes whether we have the PCF created in the constructor (and we are looping through many more "pulses") or not (and we have to create the same PCF each time. I think the main inefficiency comes with this line: `fParameterValue = opts->GetInt(detname, fDefaultParameterValues[fChannel])` since the fDefaultParameterValues and fOneSigmaValues maps only get filled once in total).

I'll try and find a way to have the correct detector name available in the generator constructor. Hopefully, I don't break anything...

**Comment by AndrewEdmonds11** [comment]
So the problem comes in MakeAnalysedPulses::AddGenerator:

``` C++
// Get the requested generator
    TVAnalysedPulseGenerator* generator=
        TAPGeneratorFactory::Instance()->createModule(generatorType,opts);
    if(!generator) return false;
    generator->SetChannel(detector);
```

we create the generator and then we set the channel after it's been created. I don't want to play with the createModule() function so, for the time being, I'll try and overload the SetChannel function to also create the PCF

**Comment by AndrewEdmonds11** [comment]
> I'll try and overload the SetChannel function to also create the PCF

Can't do this because fSource is private. This is probably for good reasons so I'll try and find another way. Instead of spamming people I'll just let you know when I solve it

**Comment by AndrewEdmonds11** [comment]
So I left this issue for a bit and tested the new production config file. A couple of small fixes:
- corrected the name of `PlotIntegral` in production.cfg
- uncommented the calculation of `integral` from the TAP generator

**Comment by benkrikler** [comment]
Good morning!  So where have we got to?

> > I'll try and overload the SetChannel function to also create the PCF
> 
> Can't do this because fSource is private.

That sounds like the right way to go.  You shouldn't need to touch fSource though, we just need to make SetChannel virtual then you call `TVAnalysedPulse::SetChannel` from the derived class' version.  I can quickly put this in now, but I don't want us to lose more time with this as this point.

**Comment by AndrewEdmonds11** [comment]
So, I ended up doing the following, which should mean we create a PCF once for each detector but it hasn't sped up the program:

In TVAnalysedPulse:

``` C++
@@ -7,6 +7,7 @@
 #include "definitions.h"
 #include "TAPGeneratorOptions.h"
 #include "TAnalysedPulse.h"
+#include "PulseCandidateFinder.h"

 class TPulseIsland;

@@ -43,6 +44,11 @@ class TVAnalysedPulseGenerator {
         /// \todo Document this
         virtual bool MayDivideTPIs()=0;

+       /// \brief
+       /// This method is overloaded by a generator if it MayDivideTPIs
+       void SetPulseCandidateFinder(PulseCandidateFinder* pcf) { fPulseCandidateFinder = pcf; }
+       PulseCandidateFinder* GetPulseCandidateFinder() { return fPulseCandidateFinder; }
+
         void SetPulseList(PulseIslandList* list){fPulseList=list;};
         /// \brief
         /// The suggested method for constructing a new TAP.
@@ -114,6 +120,10 @@ class TVAnalysedPulseGenerator {
         /// \brief
         /// The vector of TPIs being processed.
         PulseIslandList* fPulseList;
+
+       /// \brief
+       /// The pulse candidate finder that the generator may or may not use
+       PulseCandidateFinder* fPulseCandidateFinder;
 };
```

In MakeAnalysedPulses.cpp:

``` C++
@@ -207,7 +207,9 @@ bool MakeAnalysedPulses::AddGenerator(const string& detector,string generatorTyp
         TAPGeneratorFactory::Instance()->createModule(generatorType,opts);
     if(!generator) return false;
     generator->SetChannel(detector);
-
+    if (generator->MayDivideTPIs()) {
+      generator->SetPulseCandidateFinder(new PulseCandidateFinder(detector, opts));
+    }
```

and I made the appropriate changes in FirstCompleteAPGenerator.

As I said, it's made no real difference, so I'm currently profiling the program (gprof didn't work for some reason so now trying valgrind).

Not sure how you want to play this. I think it would be nice to do a production without the PCF first so we can see how it changes things later. If we run with the PCF for tomorrow, we should just do a handful of runs (we can do 20 in parallel on Merlin, right?) and then possibly run over all of them for a couple of days and save the TAPs so we save time later.

**Comment by AndrewEdmonds11** [comment]
Ok, I think I've found it. It's because in `FindCandidatePulses_Slow` and `FindCandidatePulses_Fast`, I do this:

``` C++
  double pedestal = SetupNavigator::Instance()->GetPedestal(bankname);
  double noise = SetupNavigator::Instance()->GetPedestalError(bankname);
```

which goes through the SQLite file. We already have the pedestal noises in `fOneSigmaValues` and so I'll just use those and then do something similar for the pedestals so we don't have to keep retrieving them each time.

BTW, the commands I ran to get a nice GUI for the profile:

``` bash
> valgrind --tool=callgrind ./rootana -i ......
> kcachegrind callgrind.out.X
```

**Comment by benkrikler** [comment]
Ah, sorry Andy, I was maybe a bit hasty, cause I just added the code I was suggesting above and pushed it.

I don't think however that MakeAnalysedPulses should be setting things on the generator like this.  It breaks the way we separate out the algorithms I think.  In any case, what you suggest is a bit of work-around which should no longer be needed.

> Not sure how you want to play this. I think it would be nice to do a production without the PCF first so we can see how it changes things later.

How about this then,  I'll focus on getting a production going using just MaxBin which will still be able to profile and check the plotters and TDP creation since it fills time and amplitude.  In the mean-time, you can try and speed up the PulseFinder?

**Comment by benkrikler** [comment]
Must of hit post at the same time there!  That definitely sounds like a problem.  Can you move it into the constructor and then test?

I was using `callgrind_annotate` last night, but it just gives you text output, so it's a bit trickier to read maybe

**Comment by AndrewEdmonds11** [comment]
So I replaced this:

``` C++
double noise = SetupNavigator::Instance()->GetPedestalError(bankname);
```

with this:

``` C++
double noise = fOneSigmaValues[IDs::channel(fChannel)];
```

in both the fast and slow algorithms.

This meant 5 events took 2 minutes rather than 4. Also, if I store the noise and pedestal as member variables and retrieve them once like you suggest, then it will be much quicker but we will see...

**Comment by benkrikler** [comment]
Avoid doing any sort of `find` algorithm in the main processing functions.  That includes `operator[]` on a map.  Since we know the channel from the constructor, we should initialise everything immediately.  If that uses a map or something else that's fine, but the look-up of the value in that table should not need repeating each time we're asked to process a TPI.

**Comment by benkrikler** [comment]
I've just run the production module file but using MaxBin as the only generator.  On run 3329 which has 1409 entries, it took just under 3 minutes.  Do you reckon we can squeeze the MegaTAP generator into 10 times that?  The output file size is about 91MB.  I'll add in the other old generators to compare this to the last run.

**Comment by AndrewEdmonds11** [comment]
Ok, with all my changes. Run 2808 (2609 entries) took 8 minutes :D 

So, this looks like problem solved, right? Also, I've been doing this work on your BK_prepare_production_2 branch, Ben, so shall I commit and push my changes now before you do anything else?

**Comment by benkrikler** [comment]
Great! I've been burning up resources using 3 generators for each channel (It makes an awful lot of plots), so having a functioning one which does it all is really good.

I would commit and push, but first make sure to pull, and if you're ok with it, could you remove the changes you described above to MakeAnalysedPulses and TVAnalysedPulse?  Or you can just commit and push and I'll dig around after.

We can then tag this and try running this beast...

**Comment by AndrewEdmonds11** [comment]
> could you remove the changes you described above to MakeAnalysedPulses and TVAnalysedPulse?

I was just doing this and implementing the changes you described. Here is what I have at the moment for the overloaded SetChannel function:

``` C++
virtual void SetChannel(const std::string& det){ TVAnalysedPulseGenerator::SetChannel(det); fPulseCandidateFinder =  new PulseCandidateFinder(GetChannel.str(), fOpts); };
```

and the compiler is complaining that the TVAnalysedPulseGenerator::SetChannel is also private. MakeAnalysedPulses got around it because it is a friend class. any ideas here?

**Comment by benkrikler** [comment]
You can just pull this branch cause i added the changes and pushed it all.

**Comment by AndrewEdmonds11** [comment]
Ok, so I committed my changes to PulseCandidateFinder and FirstCompleteAPGenerator and then pulled in your changes (needed to resolve a really simple conflict) and now everything is pushed.

However, it is now not compiling with the error:

```
Making object: 'obj/FirstCompleteAPGenerator.o'
In file included from src/TAP_generators/FirstCompleteAPGenerator.cpp:2:0:
src/TAP_generators/FirstCompleteAPGenerator.h:38:17: error: ‘virtual void FirstCompleteAPGenerator::SetChannel(const string&)’ cannot be overloaded
In file included from src/TAP_generators/FirstCompleteAPGenerator.cpp:2:0:
src/TAP_generators/FirstCompleteAPGenerator.h:23:17: error: with ‘virtual void FirstCompleteAPGenerator::SetChannel(const string&)’
In file included from src/TAP_generators/FirstCompleteAPGenerator.cpp:2:0:
src/TAP_generators/FirstCompleteAPGenerator.h: In member function ‘virtual void FirstCompleteAPGenerator::SetChannel(const string&)’:
src/TAP_generators/FirstCompleteAPGenerator.h:38:156: error: ‘((FirstCompleteAPGenerator*)this)->TVAnalysedPulseGenerator::GetChannel’ does not have class type
make: *** [obj/FirstCompleteAPGenerator.o] Error 1
```

which doesn't make any sense to me.

Also, on a aide note, we'll need to do a production run of PlotPedestalAndNoise to get the SQLite database for all the runs we'll do.

**Comment by AndrewEdmonds11** [comment]
Solved it. There were two versions of the same function floating around after a merge.

**Comment by benkrikler** [comment]
Cool!  So we're good to go.  You've done this before with alcapana right?  We need to merge this feature branch into develop, then take that into master then tag it all right?

**Comment by AndrewEdmonds11** [comment]
Well, the git flow way is to do a release branch, which is kind of what we've done already. I'll do it just as soon as I check rootana runs correctly

Also, how are we going to generate the SQLite file?

**Comment by benkrikler** [comment]
Oh bugger, you mean this should have been a release not a feature?  Can we make the release branch then use native git to merge the feature branch into that then use git flow to finish it all?

Do we need an sqlite file for each run or can we reuse one for a single dataset?  I guess it depends how stable the DAQ was so one per run makes more sense?  This is a one off process so future productions won't need to run this unless we change the module correct?  

@jrquirk, can you tell us how to set up your production script?  Is it able to call some generic executable / bash script or does it directly run rootana with a given configuration?

**Comment by AndrewEdmonds11** [comment]
> Oh bugger, you mean this should have been a release not a feature? Can we make the release branch then use native git to merge the feature branch into that then use git flow to finish it all?

I was just going to create a release branch and then finish it straight away. I'll make a not of the commands and add them to the wiki after I've finished.

> Do we need an sqlite file for each run or can we reuse one for a single dataset? I guess it depends how stable the DAQ was so one per run makes more sense? This is a one off process so future productions won't need to run this unless we change the module correct? 

The way SetupNavigator is coded right now is that it also has to match the run number so, for the time being, we will need one per run.

We could either have two config files in this tag (one with just PlotPedestalAndNoise) and do those first (shouldn't take too long) and then run with the proper config.

**Comment by benkrikler** [comment]
> We could either have two config files in this tag (one with just PlotPedestalAndNoise) and do those first (shouldn't take too long) and then run with the proper config.

I think that's the way we'll have to go. 

I would like to stop the SetupNavigator from just crashing though when it's missing the input sqlite file.  If we miss a run in the production or miss-name files things should throw out meaningful errors and then stop, but at the moment it just seg-faults.

**Comment by AndrewEdmonds11** [comment]
Ok, I've updated the master branch now.

One problem, I've just thought of, is that the SQLite database is created local to the user and so if we are all running PlotPedestalAndNoise on different runs we'll all end up with different versions of the table and there's no guarantee we would run the same runs on the second go-through.

Shall I just run over all the Al50(b) runs with PlotPedestalAndNoise first before we do the production run? @jrquirk, is there a way to only select certain runs to be processed in your jobscripts?

**Comment by benkrikler** [comment]
John's wiki on the production scripts is fairly complete.  Correct me if I'm wrong John, but there's not an option to run on just one dataset?  Also, I can't see where we set the task to run for this production, but given you store the modules file in the database, am I right that you call rootana directly and then supply the modules file to it?

**Comment by AndrewEdmonds11** [comment]
Ok, so I guess John is otherwise engaged. I've tried running on the Merlin login cluster for one run with PlotPedestalAndNoise only and it took two minutes.

I can set up a script to run over just the Al50b runs (there are 157 of them) which would take about 5 and a half hours. What time would it be in US/Brazil after this? I could start running now and then let you guys copy the SQLite file and do the actual production yourselves. It's a very roundabout way of doing it but if we want some results by tomorrow then this is probably the best bet.

**Comment by benkrikler** [comment]
5 hours from now is about 5pm for me.  

Why don't we divide up the runs and just process 10 or 20 in their entirety then go for the rest if there's time?

**Comment by AndrewEdmonds11** [comment]
So we wouldn't be doing this on the Merlin batch system for the time being then. Is that what you mean?

**Comment by jrquirk** [comment]
If you want to use the jobscripts, you can start a new production with the `--new` flag with something like

```
$ ./run_production.py --production=rootana --new=rootana_v19 --version=3
```

And then everyone else can just run

```
$ ./run_production.py --production=rootana
```

and they'll latch on to your rootana production version. In the above, the `--version=3` option means "base this new production off of alcapana version 3," the rootana version will be automatically incremented from the last rootana production version and will be printed out in all cases.

Or you can write your own script. I've just been using short bash scripts for my work recently just out of habit.

**Comment by jrquirk** [comment]
And there's no way as of yet to specify a dataset.

**Comment by AndrewEdmonds11** [comment]
Ok, I've just found (and fixed) a bug in PlotPedestalAndNoise where extra runs were not getting lines added to the SQLite table.

So shall we just run about 10 runs each tonight (I've got about half of the script we would need), we look at the plots tomorrow in the meeting (presumably we can just hadd the files?) and then we do the full production for Friday. How does that sound?

**Comment by benkrikler** [comment]
If we can run on the batch system that's better, but if we don't do many runs then using the login shell is probably fine.

@jrquirk, where in all this do we specify what get's run?  I found the `batch_rootana.sge` script in analyzer/batch.  Is this what is sent to the batch system then? If we want to do two different passes of a single run using 2 different modules files then how would we do it using your scripts?  Or are you suggesting we just write small one-off scripts to do this?

If we're each able to do 10 runs tonight that would be good at least to give us a check of what's working and to do an initial inspection of the results at tomorrow's meeting.

**Comment by AndrewEdmonds11** [comment]
Ok. I've merged my hotfix into master now. Here's a script I use to run several files with just the PlotPedestalAndNoise module on:

``` bash
#!/bin/bash                                                                                                                                                                                          
# generate_pedestal_and_noise.sh                                                                                                                                                                     
# -- runs rootana over a set of runs with just the PlotPedestalAndNoise module on so that                                                                                                            
#    the peedestals-and-noises.sqlite file can be created                                                                                                                                            

RUN_NUMBERS=( 3{563..564} )

for run in ${RUN_NUMBERS[@]} ; do
    filename="/gpfs/home/edmonds_a/data/tree/tree0$run.root"
    outfilename="out$run.root"
    if [ -e $filename ] ; then
        echo "$filename exists"
        ./rootana -m andyModule.txt -i $filename -o $outfilename
    fi
done
```

where andyModule.txt is:

``` cfg
#debug=all
list_modules
dump_contents

[ MODULES ]
plot_pedestal_and_noise = PlotPedestalAndNoise
## Get / save pulses
# load_pulses = LoadPulses
#analyse_pulses = MakeAnalysedPulses
#detector_pulses = MakeDetectorPulses
#SavePulses

## Plot TAPS
#PlotTime
#PlotAmplitude
#PlotIntegral

## Plot TDPs
#PlotTDPs

## Data Quality
#dq_amp = IslandAmplitude
#dq_len = IslandLength

[ analyse_pulses ]
#analyse_channels=all
#default_fast_generator=FirstComplete
#default_slow_generator=FirstComplete

[ detector_pulses]
#time_difference=1

[ load_pulses ]
#file_name=input.root

[ plot_pedestal_and_noise ]
export_sql=1

# vim: set syn=cfg:
```

I can extend the bash script to also subsequently run with the production.cfg script so that we can do a small number of runs now, if you like.

**Comment by benkrikler** [comment]
I had a very similar looking script which should work with the batch system in a parametric job:

``` bash
#! /bin/bash

run="$GE_TASK_ID"
InputDir="$HOME/data/tree"
OutputDir="$HOME/data/output"

InFile="$InputDir/tree0$run.root"
OutFile="$InputDir/out$run.root"
ModFile="$1"

RootanaCmd="$HOME/AlcapDAQ/analyzer/rootana/rootana"

# setup env
. $HOME/AlcapDAQ/thisdaq.sh

# run rootana
$RootanaCmd -m "$ModFile" -i "$InFile" -o "$OutFile"
```

And submit it like:

```
qsub -q short.q -t 3563:3574 ./run-batch.sh -- production.cfg
```

**Comment by AndrewEdmonds11** [comment]
Ok, I've generated the sqlite file for the first 10 events (do you want more?) and then I'll leave you guys to do the proper run. I don't think it will take long to run but, if it does, I can collate the results tomorrow morning (my time).

**Comment by AndrewEdmonds11** [comment]
The sqlite file is in ~edmonds_a/AlcapDAQ/analyzer/rootana/ so feel free to link against it. If you want more runs then the script is up there (a quick question: if you ran it, would it just add lines to the linked database?). Anyway, I'm off home now so let me know if anything needs doing tomorrow before the meeting.

**Comment by benkrikler** [comment]
Have you set the access permissions for us to be able to write to that file?

**Comment by benkrikler** [comment]
I've been able to do all of the runs fr Al50b quite quickly with the batch system,  but sqlite locks the database so the values cant be written to it if we run in parallel...

**Comment by jrquirk** [comment]
The pedestal/noise database?

**Comment by benkrikler** [comment]
> The pedestal/noise database?

Yes.  I'm looking through the root documentation to see if there's anything we can do, but I've no experience with this so if you do and could help me out that would be great.  In the mean time I'll just start running in serial...

**Comment by benkrikler** [comment]
Is there a way I can make many sqlite files and then combine them in a single command?  Or if I make a simple CSV file for each run, is there a nice way to translate this into sqlite?

**Comment by benkrikler** [comment]
Ok, so I took out the TSQLite stuff from PlotPedestalAndNoises and had it just dump out a text line with some obvious prefix.  Then grep through all the outputs to build a csv file and finally used the following commands in the sqlite3 terminal:

```
create table pedestals_and_noises (run integer, channel string, bank string, pedestal float, noise float);
.separator ","
.import pedestal_noise.csv pedestals_and_noises
.backup pedestals-and-noise.sqlite
```

This method let me do all runs in parallel so in total took about 10 minutes to do the main Al50b tranche.  The file is at `~krikler_b/AlcapDAQ/analyzer/rootana/pedestals-and-noise.sqlite` on merlin.

**Comment by benkrikler** [comment]
I've been able to process runs 3563 to 3646 in about 15 minutes total.  Each file is about 75MB and when I hadd all of them together and drop the TAP_tree we get a total file size of about 1MB (the TAP_tree itself is about 2GB).  The plots need tweaks themselves, especially ones where we were auto-scaling the axes, hadd struggles to superpose these ones.  I'll see what I can achieve before tomorrow's meeting.

**Comment by benkrikler** [comment]
And if anyone is interested in looking at the output, the files are in `~krikler_b/data/out/rootana_v2-main/`.

**Comment by AndrewEdmonds11** [comment]
Cool. I'm guessing you guys are asleep so I'll try and extract the plots and make some sort of picture book from this that we can look at in the meeting later.

**Comment by AndrewEdmonds11** [comment]
Quickly looking through Ben's all.root file:
- integral plots have far fewer entries than amplitude plots
  - e.g. Ge-F: 2264 --> 3553046
  - might be one of the auto-scaled plots that Ben mentioned
- some amplitude plots have a large number of entries in the bin with x-value -1 (hidden normally)
  - affects all Si slow channels, muSc, Ge-F, ScL, ScR, ScVe

Then I looked at individual runs to see if these issues arise there:
- noticed that some of the individual runs didn't close properly (I counted 33/84)
  - did you notice anything when you ran, Ben?
- of the two I checked (runs 3569 and 3575):
  - none of the amplitude plots had the large number of entries in the -1 bin
  - I'm guessing it's caused by hadding the unclosed files
- run 3569 has 2264 entries in Ge-F_Integrals and run 3575 has a different x-scale for this plot
  - so that's why the integral plots have far fewer entries than the amplitude plots in all.root

So from all this:
- `PlotIntegral` should be changed to not auto-scale the x-axis
- I'll extract the plots we do have (amplitude, time, tdiffs) for discussion later

**Comment by AndrewEdmonds11** [comment]
Looking at the PlotTDP plots, I think there's something strange going on. Some of the amplitudes being plotted are crazy (O(10^5) and/or negative):

![sir2_fastslowamplitudes](https://cloud.githubusercontent.com/assets/1848575/3824830/0347adb2-1d4d-11e4-8e40-afde390351b7.png)

Are these plots also auto-scaled, Ben? Is it that we are getting the odd crazy pulse and that we should really be zoomed in on that red spot in the top left?

PS The Ge one looks quite nice though:
![ge_fastslowamplitudes](https://cloud.githubusercontent.com/assets/1848575/3824861/5e4405d0-1d4d-11e4-9ffe-d81a4af6cc50.png)

**Comment by benkrikler** [comment]
Good morning!  Thanks for looking at things Andy, have you got anywhere with a picture-book macro?

> Some of the amplitudes being plotted are crazy (O(10^5) and/or negative):

I'm not sure what this is.  When I saw it I told myself it was because for some runs those channels had been moved to another bank, and when we do hadd the plots end up overlaid like that.  The TDP plots are all auto-scaled.

> integral plots have far fewer entries than amplitude plots 
> That's a bit alarming.  I used a different processing loop for the integrals than to the amplitude plots which were written in the early rootana days.  The integral plots are being auto-scaled, but the amplitude plots, I think, are not.
> 
> some amplitude plots have a large number of entries in the bin with x-value -1

Maybe we should allow auto-scaling? We need to refine the axes values I suppose. A mechanism similar to your default values for the pulse finder might be needed.

> noticed that some of the individual runs didn't close properly

I saw some of the files were smaller in size, but didn't pay much attention.  I grepped the output for 'error' but found nothing so assumed all was healthy.  I should definitely have checked for 'bad-alloc' though, but it slipped my mind at the time.  I'll look into this now...

**Comment by benkrikler** [comment]
@AndrewEdmonds11, Just saw your Elog.  Nice work!

**Comment by litchfld** [comment]
> Ok, so I took out the TSQLite stuff from PlotPedestalAndNoises and had it just dump out a text line with some obvious prefix. Then grep through all the outputs to build a csv file and finally used the following commands in the sqlite3 terminal [...]

You can concatenate databases natively too. See this [SO thread](http://stackoverflow.com/questions/80801/how-can-i-merge-many-sqlite-databases), for example.  

In the long run the normal way to do this is for someone (may change over time) to do this in advance of a production and 'upload' to the database, which in our case almost certainly means adding it to the repository (OK, 'cause it's a _tiny_ DB).

**Comment by benkrikler** [comment]
> You can concatenate databases natively too. See this SO thread, for example.

Not in parallel, and root's wrapping of sqlite doesn't provide much to handle situations where the DB is locked by another process, so I had to go this round about way to run things on the batch system.

**Comment by AndrewEdmonds11** [comment]
> have you got anywhere with a picture-book macro?

Not yet. I currently have a macro that prints all the plots from a TDirectory onto separate canvases and then you can go through and change/save things that you want. It's not automated.

> When I saw it I told myself it was because for some runs those channels had been moved to another bank

I think by the time we did Al50(b), channels and banks were pretty stable to I don't think this will be the case. Another possibility is that there's something going on with the pointers to our TAPs but I don't know enough about that side of the framework to judge whether this is likely.

> Maybe we should allow auto-scaling?

I'd rather not have this because of the problems we've seen where we can't easily hadd histograms together.

**Comment by benkrikler** [comment]
> Another possibility is that there's something going on with the pointers to our TAPs but I don't know enough about that side of the framework to judge whether this is likely.

What makes you suspect this and in what way are you thinking they're going wrong?

> I'd rather not have this because of the problems we've seen where we can't easily hadd histograms together.

Makes total sense.  We'll just have to inspect the overflow / underflow bins I guess.

**Comment by AndrewEdmonds11** [comment]
> What makes you suspect this and in what way are you thinking they're going wrong?

It's just some bit of junk that's at that memory address. But it does seem weird that Ge is fine.

Maybe it's a problem with the pulses we get from the PulseCandidateFinder. I'll try and have a quick look into this before the meeting.

**Comment by benkrikler** [comment]
> It's just some bit of junk that's at that memory address

That would suggest the TAP fields are not initialised correctly, but the constructors of TAPs set them to `definitions::DefaultValue` which is currently -999999.  Since we don't see a single big spike (the plot above shows at least 2 different bins are filled) at -99999 in x and y, I'm not sure this is the issue and in any case, the `MaxTimeDiffDPGenerator` is supposed to avoid creating TDPs.  If either of the TAPs have an `amplitude == defintions::DefaultValue` the TDP get's plotted in the unpaired 1D plots.

**Comment by litchfld** [comment]
I'm suspicious there might be a +/- 9999 in there somewhere.  If you get a reasonable value, and add/subtract a big 'default' number instead of a correct offset you typically end up with a distribution that has a similar spread to the correct distribution, but offset by approximately _big_default - average_correction_. Which is what we see.  In contrast memory junk usually produces a much wider spread, and in cases where there is a preferred value, the offset is unlikely to be almost exactly 10^5, as it is here.

**Comment by AndrewEdmonds11** [comment]
So I've checked in run 3569 for TAPs with amplitude < 0 by using `ExportPulse` in `PlotAmplitude` and there are none. Also the 2D plot looks fine. I'll check a few more runs to see if it's just one dodgy run.

Also, could it be from the runs that failed? Do any of them have partially completed histograms with dodgy values arising from a +-999999 that eventually caused a crash?

**Comment by AndrewEdmonds11** [comment]
Ok. I ran on the next run (3570), which is one of the ones which failed for Ben yesterday and I'm getting loads and loads of these errors:

```
Skipping out of range pulse: 605 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 606 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 607 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 608 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 609 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 610 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 611 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 612 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 613 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 614 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 617 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 618 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 619 on detector 'muSc#FirstComplete#any
Skipping out of range pulse: 621 on detector 'muSc#FirstComplete#any
```

and the final PlotTDP plot that's produced?

![problem](https://cloud.githubusercontent.com/assets/1848575/3827499/74b1a1ee-1d6e-11e4-82eb-a7375393d5ba.png)

Does anyone know where the "Skipping out of range pulse" error comes from?

**Comment by AndrewEdmonds11** [comment]
> Does anyone know where the "Skipping out of range pulse" error comes from?

grep does. It's the ExportPulse module.

**Comment by benkrikler** [comment]
I've gone through all the individual runs and got the SiR2 histograms for the TDP amplitudes.  There are 51 plots in total, so that's all the ones that didn't fail.  It seems like there are 2 distinct classes of plots.  Ones which look nice and begin around the origin and ones which come out with amplitudes +- O(10^5).

## Good results:

![run3576_sir2_tdp_amps](https://cloud.githubusercontent.com/assets/4083697/3827697/58c77cf4-1d70-11e4-9141-e38d50a36b1a.png)

## Bad results:

![run3587_sir2_tdp_amps](https://cloud.githubusercontent.com/assets/4083697/3827704/609697ee-1d70-11e4-8913-5e340510b3d0.png)

**Comment by AndrewEdmonds11** [comment]
So I've solved the problem I had run run 3570. It was because I didn't have the pedestal and noise values for that run in my local SQLite database.

After adding them, the "Skipping out of range pulse" errors have gone and the plot looks fine.

Could this be the problem from your run last night, Ben?

**Comment by benkrikler** [comment]
Bingo!  My pedestals and noise file contains 50 runs, so that sounds like that's the problem.   I guess PlotPedestalAndNoise crashed and I didn't notice it?  Indeed, when I scan through each run's plot, those runs that don't appear in the table are the ones with odd distributions.

The runs that failed were due to bad_allocs, but it perhaps the problems are related. Maybe we have an extra leak if the data doesn't exist in the sqlite db?

**Comment by benkrikler** [comment]
Didnt mean to close this, sloppy mouse work on my end...

Btw, is anyone else having issues with the batch system?

**Comment by AndrewEdmonds11** [comment]
Ok, from the meeting today the things we need to do are (**very important**):
- **add a parameter value to the PulseCandidateFinder for NDet2**
- **generate pedestal and noise SQLite database** (either serially or parallelally)
- **change axis limits on PlotTime module** (to be 0 - 110ms rather than 0 - 100ms)
- validate SQLite database
- add John's module for time differences between fast and slow pulses
- allow jobscripts to run over a single dataset (and also run twice with two different modules files)

I will make a new release branch for the little changes. I think if anything fairly large gets added we can just rebase the release branch on develop.

**Comment by benkrikler** [comment]
We also want to:
- Fix the axes for TDP plots and title labels
- Merge in Damien's TDiff module (once it's ready) and profile it in a run

**Comment by AndrewEdmonds11** [comment]
I've done the parameter value and changed the axis limits on `release/rootana_v2.1`. I'm off home now so feel free to use this branch for the little changes.

also, when do we actually have to do the production so that we have things ready? I can compile things while you're all asleep on Thursday night, if you start the production when you leave work.

Let me know what needs looking at before then.

**Comment by jrquirk** [comment]
I've set up two new tables for production. First I created two new tables with all datasets by running the following command twice

```
$ ./run_production.py --production=rootana --new=rootana_v2 --version=3
```

The value of `--new` is the repository tag. This created two new versions of rootana in the production database, versions 2 and 3 (both using rootana_v2 tag of the code, but having table names in the production database rootana_v2 and rootana_v3; it's unnecessary to point out the flaw here). I then removed all runs not associated with datasets other than the Aluminum 50 B

```
$ for r in $(sqlite3 ~/data/production.db "SELECT run FROM datasets WHERE dataset != 'Al50b'"); do
>     sqlite3 ~/data/production.db "DELETE FROM rootana_v2 WHERE run==$r";
>     sqlite3 ~/data/production.db "DELETE FROM rootana_v3 WHERE run==$r";
> done;
```

Now, to run for pedestals/noise, just make sure you have the correct AlcapDAQ/analyzer/rootana/production.cfg file and run

```
$ ./run_production.py --production=rootana --version=2
```

And then, when that's done, replace the production.cfg file with the new one you want and run

```
$ ./run_production.py --production=rootana --version=3
```

Remember the files get saved to data/out/v2/ and data/out/v3/, and the most recently run production gets linked in data/out/. Specifically @benkrikler, who mentioned he was going to make links; running the production scripts will change some of them around and possibly pollute your data/out directory.

**Comment by benkrikler** [comment]
Thanks John.  I realised after the meeting why what I was suggesting was a bad idea, it just seemed convenient for scripting for loops over files.

So I taught myself some sqlite and found that the following command was able to give me all the tree files for a given dataset:

```
select a.run, tree from alcapana_V3 as a join datasets as d where dataset="Al50b" and a.run=d.run;
```

I've been unable to submit jobs to the Batch system this morning, they just seem to get stuck in the queue for hours or else flagged with a queue error. Is anyone else seeing this?  As a result though, I'm running over all Al50b runs in serial to generate the noise and pedestal database.  Could others try to generate the pedestal and noise db as well?

That's great that we can easily run now for a given dataset though.  I think Damien's finished the t_diff module and it just needs merging, I've put the corrections that were mentioned at the meeting into PlotTDPs.  If you merge your module in I think we've got everything we need to run.  We could in principle get this going by tomorrow then, depending on the pedestals and noise database.

The other tasks we needed to do was get a few pulses together out of the odd looking channels / distributions.  The PulseViewer module can do this easily by just putting:

```
pv=PulseViewer(ScR , amplitude > 200 )
```

**Comment by benkrikler** [comment]
**Pedestal and noise DB**
Pedestal and noise database is ready for Al50b.  Scanning through it by eye it looks like all runs are there.  All output files from each run have the same output file size.  The file is at `~krikler_b/data/pedestals-and-noises.sqlite` if you wish to soft-link against it.

**PulseViewer**
I've also extended the pulse viewer module so it can take any set of boolean logic on TAP amplitudes, times, number of samples, integral value, and pedestal by using Root's TFormula.  There are also extra options to set a limit to the number of pulses plotted.  

I've put a config file specific to using this module at `rootana/configurations/pulse_viewer.cfg`, which should be enough to show someone how to use it.  This work is pushed to develop already, so perhaps @AndrewEdmonds11 when you get in you could look into getting pulses in the dodgy areas of the plots we saw today?

**Production run**
I'll be away during the day tomorrow, but I'll be able to focus on things in the evening.  I think we've got everything we need to do the run, but I'm still having trouble with the batch system.  I've set a local serial job going for what it's worth. 

@dmalexan I see you've closed issue #169, did you merge things into develop?  If so I think you may need to push your branch for us to see the changes; git flow doesn't do that for you.

**Comment by AndrewEdmonds11** [comment]
Yes, I can look into the dodgy areas. I'll open a new issue for this.

**Comment by AndrewEdmonds11** [comment]
For production where should the SQLite database be?

When I initially coded for it, it needed to be in the same directory as you were running rootana. We will probably want to change this for production. Just a note for later.

**Comment by benkrikler** [comment]
I looked into rebasing the release branch.  It seems git flow doesn't allow us to do that.  Do we want to finish the 2.1 release then open a new one to bring in changes from develop? 

Also, I've been running the current 2.1 release overnight (in serial since I've not been able to use the batch system) and have currently gone through 120 runs without a single one throwing bad_alloc.  My guess is our interactive terminals have more memory available than the batch system then.

> For production where should the SQLite database be?

I think it's ok as it is.  We can always make symbolic links to another file.  An alternative is to provide a command line argument to control the file's location or add an option to the modules file.

**Comment by AndrewEdmonds11** [comment]
> I looked into rebasing the release branch. It seems git flow doesn't allow us to do that. Do we want to finish the 2.1 release then open a new one to bring in changes from develop? 

Yes, I think that's probably the best way.

**Comment by benkrikler** [comment]
I've hadd-ed the runs which have currently finished from v2.1.  The file is at: `~krikler_b/data/out/rootana_v2-Al50b/all3563_3702.root`.  I've been through quickly and things look much better from the TDPs.  The integral distributions look interesting as well.  The TDiff plots from TDPs show some instability in the relative timing between fast and slow.  I put these comments [elog:157](https://muon.npl.washington.edu/elog/mu2e/Analysis-R13/157?suppress=1) and will try to upload images there as well before I need to leave.

**Comment by thnam** [comment]
It might not be relevant now, but should we hadd histograms at this stage? Or should we wait after calibrating?

**Comment by benkrikler** [comment]
Why do you think that?  I feel that calibration will only change our axes labels to show energy rather than raw amplitude or integral, but otherwise many of the distributions we're interested in will look the same, no?

**Comment by thnam** [comment]
I'm just worried if there is some drifting over time. Probably the effect would not be too large in this case.
Also I imagine that the main products of this rootana run would be analysed pulses and detector pulses that will be used to form muon events later. The histograms at this stage are mainly for sanity check at a (little?) higher level, so it would be more convenient to have one histogram for one run.

**Comment by litchfld** [comment]
If it is possible to calibrate on a run-by run basis then I think we will want to do it, eventually.  But for now lets just assume it isn't time dependent and press on!  

Having said that, once we have a reasonable selection for real physics pulses we should absolutely make time-evolution histograms, like in the previous picture books.  Those distributions were in some respects uninformative because the pulses we care about were normally drowned out by the noise pulses, so redoing them without noise is definitely a useful analysis.

**Comment by litchfld** [comment]
> I think it's ok as it is. We can always make symbolic links to another file. An alternative is to provide a command line argument to control the file's location or add an option to the modules file.

Symbolic links are terrible from the perspective of keeping a paper trail. (it's not so bad in the case where the filename is preserved, but swapping data in/out by changing a link is bad). Much better to have the file named explicitly in the code or options.

**Comment by benkrikler** [comment]
> Symbolic links are terrible from the perspective of keeping a paper trail.

The run db implemented by John stores a copy of the contents of the modules file that was used, plus we've been running with `dump_contents` which dumps out the module file, plus many of the configurations get stored as strings in the root file itself.  

On the other hand, a hard-coded file location makes it awkward to change what modules file we run with (the soft-link suggestion is a hack to get around exactly this issue) whereas a run-time option doesn't do much better since I could just as easily then change or move elsewhere the file that was used.

From the paper trail perspective, the only good solution is to store the modules file alongside, or better still inside, the data.

**Comment by benkrikler** [comment]
I think this issue has become big and unwieldy enough.  I'm going to close this and open a new one for the next production run based on today's discussion.

**Comment by litchfld** [comment]
Re: symlinks. If the file is dumped that makes the point moot in this instance  [_I'm not sure it is though, we are talking about the calibration DB, which are not part of the modules file_] but we can only get away with that because of our low channel count.  In general though symlinks should be studiously avoided.

> [...] On the other hand, a hard-coded file location makes it awkward to change what modules file we run with (the soft-link suggestion is a hack to get around exactly this issue) [...]

If you need to change it multiple times during a production then a softlink is a particularly foolish approach, since it can interfere with any other process running at the same time.  But if you are going to keep the same file all production, hardcoding is not really a problem since recompilation is a negligible cost.

> [...] whereas a run-time option doesn't do much better since I could just as easily then change or move elsewhere the file that was used. [...]

Yes but then that's either an embarassing accident or deliberate stupidity. Whereas using a softlink erasing the paper trail is _required_ by the process.

So as a general principle keep well away from symlinks for configuration data, unless the program is specifically set up to interrogate them.  Consider (schematically):

``` C++
// Configuration data file, is a softlink:
std::string cfg_filename = "../config/config-db.sqlite";
ReadConfig(cfg_filename);
RecordConfig(cfg_filename);  //writes "../config/config-db.sqlite" to the output file
```

v.s.

``` C++
// In practice it may be easier to set this at run time, depending on how often it changes  
std::string cfg_filename = "../config/production3_andy_tweakedfits_20140811.sqlite"
ReadConfig(cfg_filename);
RecordConfig(cfg_filename);  //writes "../config/production3_andy_tweakedfits_20140811.sqlite"
```

Even if we are dumb enough to not archive the file, and fat-fingered enough to delete it completely, the latter case still gives us an idea where the data came from, if we discover a problem down the line. ("Ah, it was made by Andy on 11th of August.  According to my emails that was before we discovered some nasty bias in the silicon channels. We should have changed that before we respun... guess we have to do it again <sigh>"). In the former case you have no idea, you either have to cross your fingers and hope, or redo everything at the slightest suspicion of a problem.

**Comment by jrquirk** [comment]
I've removed the automatic creation of symlinks in the job scripts. Previously the job scripts would create symlinks to the most recently produced output file (data/out/out<run>.root -> data/out/v<version>/out<run>.root). I know that isn't what's being discussed here, but this issue motivated that change.

Given the separation between the run scripts, the command line arguments, and the calibration DB being loaded in from a module, can the  constructor of the module used to load the database dump the contents to standard output so it's saved in the log? As Phil said, there aren't many channels so it wouldn't clog up the output.

**Comment by AndrewEdmonds11** [comment]
> can the constructor of the module used to load the database dump the contents to standard output so it's saved in the log?

Are we still talking about (what is currently called) pedestals-and-noises.sqlite? If so, it's currently done in SetupNavigator and I was going to start working on this to make it a bit more user friendly so will add this in. (Also, some modules I've written interact with the SQLite database which I guess we shouldn't have (see issue #137)).

**Comment by AndrewEdmonds11** [comment]
No, I didn't sorry. Is it not included from our group permissions?

**Comment by jrquirk** [comment]
The group write bit will take care of that.

**Comment by AndrewEdmonds11** [comment]
Well, I'm pretty I sent that last message ages ago. Must have got held up in my phone's outbox...
